src/arch/io/network.cc

```diff
@@ -621,6 +621,430 @@ void linux_tcp_conn_t::on_event(int /* events */) {
+                probably signalled by on_shutdown(). */
+                throw tcp_conn_read_closed_exc_t();
+            }
+            /* Go around the loop and try to read again */
+            continue;
+        case SSL_ERROR_WANT_WRITE:
+            /* Though we are reading, a TLS renegatiotain may occur at any time
+            requiring a write. Wait for a notification from the event queue, or
+            for an order to shut down */
+            {
+                linux_event_watcher_t::watch_t watch(get_event_watcher(), poll_event_out);
+                wait_any_t waiter(&watch, &closed);
+                waiter.wait_lazily_unordered();
+            }
+            if (closed.is_pulsed()) {
+                /* We were closed for whatever reason. Whatever signalled
+                us has already called on_shutdown(). */
+                throw tcp_conn_read_closed_exc_t();
+            }
+            /* Go around the loop and try to write again */
+            continue;
+        case SSL_ERROR_SYSCALL:
+            // An error with the underlying IO.
+            io_err = ERR_get_error();
+            if (io_err != 0) {
+                // The Openssl error queue has data on the error.
+                on_shutdown();
+                return;
+            }
+
+            // if ret is 0 that indicates an EOF in violation of the protocol.
+            if (0 == ret) {
+                on_shutdown();
+                return;
+            }
+
+            // Otherwise consult errno.
+            switch (get_errno()) {
+            case EPIPE: case EHOSTUNREACH: case ENETDOWN:
+            case EHOSTDOWN: case ECONNRESET: case ENOTCONN:
+                /* These errors are expected to happen at some point in practice */
+                on_shutdown();
+                return;
+            default:
+                /* Unknown error. This is not expected, but it will probably happen sometime so we
+                shouldn't crash. */
+                on_shutdown();
+                return;
+            }
+        default:
+            // Some other error.
+            on_shutdown();
+            return;
+        }
+    }
+}
+
+void linux_secure_tcp_conn_t::shutdown() {
+    assert_thread();
+
+    int ret;
+
+    // If we have not already received a "close notify" alert from the peer,
+    // then we should send one. If we have received one, this loop will respond
+    // with our own "close notify" alert.
+    while (true) {
+        ERR_clear_error();
+
+        if ((ret = SSL_shutdown(conn)) >= 0) {
+            break; // "close notify" was either sent or received.
+        }
+
+        // A negative ret indicates as error.
+        int err_code = SSL_get_error(conn, ret);
+        if (SSL_ERROR_WANT_READ == err_code) {
+            linux_event_watcher_t::watch_t watch(get_event_watcher(), poll_event_in);
+            wait_any_t waiter(&watch);
+            waiter.wait_lazily_unordered();
```

Comment:
This is a bit risky because if the remote side never reads from the socket or (further down) doesn't send a response, we will be stuck in here forever waiting for `waiter` I think. The server could not even be shut down in that case.
I suggest adding a timeout to `waiter` for all waits in this `shutdown()` function. After the wait you can check if the timeout was pulsed. If it was, we continue streight to line 1024 and simply close the socket.

I suggest using a single timeout for the whole `shutdown()` process, rather than one per wait.

Something like this:
```
// <at the beginning of `shutdown()`:>
// Wait at most 5 seconds for the orderly shutdown. If it doesn't complete by then, we
// simply shutdown the socket.
signal_timer_t shutdown_timeout(5000);

// <further down in the waiter code:>
wait_any_t waiter(&watch, &timeout);
waiter.wait_lazily_unordered();
if (timeout.is_pulsed()) {
  break;
}
```